### PR TITLE
throw an exception on id.get if session is not available

### PR DESCRIPTION
**Pull request by iamkrillin**
Throw exception on id.get
- [ ] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [ ] You've included unit or integration tests for your change, where applicable.
- [ ] You've included inline docs for your change, where applicable.
- [ ] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Summary of the changes (Less than 80 chars)

## Description
Throw an exception in DistributedSession.Id.Get if the session is not available

### CODE REVIEW

**Comment by amcasey** [comment]
It [looks](https://github.com/dotnet/aspnetcore/blob/bbe63a6c5c548a076062502cde405fb2c8d70462/src/Middleware/Session/src/DistributedSession.cs#L199) like the unavailability of a session is already signalled by an empty `Id`.  Is that not sufficient?

**Comment by iamkrillin** [comment]
The problem I was running into was two fold:

1. The documentation says the Id returned is unique, which strictly speaking, if there is a possibility for it to return "" is not always true, but the bigger problem was..
2. When using sqlserverdistributedcache, if there is a problem with the setup (server not available, table missing, etc).  The error gets eaten then logged to the console (which is not ideal when using IIS hosting)

These two things combined lead to a scenario where things looked to be working, but were not working.  When I was working on this PR I though it was better to throw an exception since that would make it _very_ clear that things were not working, but returning null would also work

Either of these would be preferable to eating a error, then returning a value IMO

**Comment by amcasey** [comment]
I'm not actually familiar with this type, but the current behavior certainly appears to be intentional.  If we were going to move to a model where it throws, wouldn't it make more sense to just add a `throw` in `Load`s catch block?

Similarly, it would have been easy to give each failed session its own guid, so I assume they were grouped together with the id `""` on purpose.  (Whether that purpose is valid, I have no idea, though it's a common enough pattern.)

**Comment by iamkrillin** [comment]
That's a fair enough point, I did consider adding a throw into the catch in Load, I ultimately decided against it since Load is called in many places and I wanted my change to be as small as possible.

**Comment by amcasey** [comment]
I was hoping an expert would chime in, but I'm not sure there is one for this type. 😆 

Barring that, I think I'd prefer to just set it to `null`, rather than `string.Empty` in `Load`, which should cause it to use a fresh guid.  Having said that, I can also see the value in just leaving it.  If there _is_ a session ID and it just couldn't be loaded, assigning a fresh one is a little strange.  It looks like the expectation of the API is that the other properties will be disregarded if `IsAvailable` is false.

If you need more explicit error behavior in your app, do you have the option of either checking `IsAvailable` or switching to `LoadAsync`, which appears to throw when you want it to?

**Comment by iamkrillin** [comment]
I changed my application to check IsAvailable before opening this pull request, but figured if this behavior impacted me, perhaps it is impacting others.  My Application was checking for null before as an error condition so that makes me think at some point in the past perhaps this API returned null in that scenario. My application has been around since .net 3 and is being updated to .net core so it may have been _long_ ago.

File: src/Middleware/Session/src/DistributedSession.cs [hunk 1]
```diff
@@ -96,6 +96,10 @@ public string Id
         get
         {
             Load();
+            if(IsAvailable == false)
+            {
+                throw new InvalidOperationException("Session not currently available");
+            }
```
Comment (mkArtakMSFT) [inline src/Middleware/Session/src/DistributedSession.cs:102]: The IsAvailable getter already calls `Load()` internally.
If the plan is to leave this check in place, let's remove the `Load()` from the line above.

**Review by mkArtakMSFT — APPROVED** [review]
Thanks for your patience.
It looks like the conversation with @amcasey landed on potentially documenting the current behavior, instead of adding this check. However, I can see the value of being explicit and given where we are in the release I wonder if it's safe to take this change. So, no objection from me, but someone who is more familiar with the codebase should share their opinion too.

**Comment by mkArtakMSFT** [comment]
@JamesNK are you by any chance familiar with this code?

**Comment by JamesNK** [comment]
> @JamesNK are you by any chance familiar with this code?

No, never used it.

IMO, I don't think we should change the behavior of the property. We could break people who depend on the current behavior. If we did change the current behavior, then I think an option would be required to revert it if needed. It starts to become a lot of work for not a huge benefit. And, I'm not sure we fully understand why this choice was made in the first place, which makes me hesitant to change it.

Instead of this change, I think we should update the documentation to say an `Id` is empty if there isn't a session and `IsAvailable` should be checked. Basically, I agree with what @amcasey said [here](https://github.com/dotnet/aspnetcore/pull/57902#issuecomment-2397951448).

**Comment by mkArtakMSFT** [comment]
Thanks @JamesNK.
Appreciate your input.

> Instead of this change, I think we should update the documentation to say an Id is empty if there isn't a session. Either LoadAsync should be explicitly called or IsAvailable should be checked

@iamkrillin would you like to update your PR to do this instead?

**Comment by mkArtakMSFT** [comment]
Actually, I'm going to close this for now. @iamkrillin feel free to reopen if you'd like to repurpose this for the documentation change.

**Pull request closed without merging**